enaml/testing/utils.py

```diff
@@ -0,0 +1,163 @@
+
+    This should be used by all tests creating windows in a teardown step.
+
+    """
+    app.process_events()
+    sleep(0.1)
+    while Window.windows:
+        for window in list(Window.windows):
+            window.close()
+        app.process_events()
+        sleep(0.02)
+
+
+class ScheduledClosing(object):
+    """Scheduled closing of dialog.
+
+    """
+
+    def __init__(self, app, cls, handler, op):
+        self.called = False
+        self.cls = cls
+        self.custom = handler
+        self.op = op
+        self.app = app
+
+    def __call__(self):
+        i = 0
+        while True:
+            dial = get_window(self.cls)
+            if dial is not None:
+                break
+            elif i > 10:
+                raise Exception('Dialog timeout')
+            sleep(0.1)
+            i += 1
+
+        try:
+            self.custom(dial)
+        finally:
+            self.app.process_events()
+            from .fixtures import DIALOG_SLEEP
+            sleep(DIALOG_SLEEP)
+            getattr(dial, self.op)()
+            self.called = True
+
+
+@contextmanager
+def handle_dialog(app, op='accept', handler=lambda x: x, cls=Dialog, time=100,
+                  skip_answer=False):
+    """Automatically close a dialog opened during the context.
+
+    Parameters
+    ----------
+    op : {'accept', 'reject'}, optional
+        Whether to accept or reject the dialog.
+
+    handler : callable, optional
+        Callable taking as only argument the dialog, called before accepting
+        or rejecting the dialog.
+
+    cls : type, optional
+        Dialog class to identify.
+
+    time : float, optional
+        Time to wait before handling the dialog in ms.
+
+    skip_answer : bool, optional
+        Skip answering to the dialog. If this is True the handler should handle
+        the answer itself.
+
+    """
+    sch = ScheduledClosing(app, cls, handler, op)
+    timed_call(time, sch)
+    try:
+        yield
+    except Exception:
+        raise
+    else:
+        while not sch.called:
+            app.process_events()
+
+
+@contextmanager
+def handle_question(app, answer):
+    """Handle question dialog.
+
+    """
+    def answer_question(dial):
+        """Mark the right button as clicked.
+
+        """
+        dial.buttons[0 if answer == 'yes' else 1].was_clicked = True
+
+    with handle_dialog('accept' if answer == 'yes' else 'reject',
+                       custom=answer_question, cls=MessageBox):
+        yield
+
+
+def show_widget(app, widget):
+    """Show a widget in a window
+
+    """
+    win = Window()
+    win.insert_children(None, [widget])
+    win.show()
+    app.process_events()
+    return win
+
+
+def show_and_close_widget(app, widget):
+    """Show a widget in a window and then close it.
+
+    """
+    from .fixtures import DIALOG_SLEEP
+    win = show_widget(widget)
+    sleep(DIALOG_SLEEP)
+    win.close()
+    app.process_events()
+
+    close_all_windows(app)
```

Comment:
Lots of `sleep` and `process_events` in this file. I wonder if a decorator or context manager would be a better approach since it would be explicit and local to the test where it is needed.

Reply:
The last two functions are a left over, we do not need them so I will remove them. I think it make sense to turn close_all_windows into a context manager. I will do so and remove the associated fixture.

Reply:
If windows_closing disappears, we can manually manage what app is passed to the test, which is cleaner in a sense even if we have only one backend currently. I will refactor the code to skip tests if qt is not available.